File: recognizer.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <float.h>

#define MAX_FRAMES 2000
#define MAX_DIMENSION 39
#define MAX_TEMPLATES 15
#define INF 1e9

typedef struct {
    int frames;
    int dimension;
    double **data;
    char label[10];
} MFCC;

MFCC *read_mfcc(const char *filename, const char *label) {
    FILE *fp = fopen(filename, "r");
    if (!fp) {
        printf("Error opening file: %s\n", filename);
        return NULL;
    }

    int frames, dimension;
    if (fscanf(fp, "%d %d", &frames, &dimension) != 2) {
        printf("Error reading header from %s\n", filename);
        fclose(fp);
        return NULL;
    }

    MFCC *mfcc = (MFCC *)malloc(sizeof(MFCC));
    mfcc->frames = frames;
    mfcc->dimension = dimension;
    strncpy(mfcc->label, label, 9);
    mfcc->label[9] = '\0';

    mfcc->data = (double **)malloc(frames * sizeof(double *));
    for (int i = 0; i < frames; i++) {
        mfcc->data[i] = (double *)malloc(dimension * sizeof(double));
        for (int j = 0; j < dimension; j++) {
            fscanf(fp, "%lf", &mfcc->data[i][j]);
        }
    }

    fclose(fp);
    return mfcc;
}

double calculate_distance(double *v1, double *v2, int dim) {
    double sum = 0.0;
    for (int i = 0; i < dim; i++) {
        double diff = v1[i] - v2[i];
        sum += diff * diff;
    }
    return sqrt(sum);
}

// One-Pass DP for Continuous Digit Recognition
void recognize(MFCC *test, MFCC **templates, int num_templates) {
    int T = test->frames;
    
    // best_score[t]: Minimum cost to reach time t having completed a word
    double *best_score = (double *)malloc(T * sizeof(double));
    // backptr_word[t]: Which word ended at time t
    int *backptr_word = (int *)malloc(T * sizeof(int));
    // backptr_time[t]: When did that word start (or previous word end)
    int *backptr_time = (int *)malloc(T * sizeof(int));

    for (int t = 0; t < T; t++) {
        best_score[t] = INF;
        backptr_word[t] = -1;
        backptr_time[t] = -1;
    }

    // Internal DP state for each template: current_cost[template][template_frame]
    double **current_cost = (double **)malloc(num_templates * sizeof(double *));
    double **next_cost = (double **)malloc(num_templates * sizeof(double *));

    for (int k = 0; k < num_templates; k++) {
        int ref_frames = templates[k]->frames;
        current_cost[k] = (double *)malloc(ref_frames * sizeof(double));
        next_cost[k] = (double *)malloc(ref_frames * sizeof(double));
        for (int j = 0; j < ref_frames; j++) {
            current_cost[k][j] = INF;
        }
    }

    // Initialization: At time -1, score is 0
    double prev_best_score = 0.0;

    for (int t = 0; t < T; t++) {
        // For each template
        for (int k = 0; k < num_templates; k++) {
            int ref_frames = templates[k]->frames;
            
            for (int j = 0; j < ref_frames; j++) {
                double dist = calculate_distance(test->data[t], templates[k]->data[j], test->dimension);
                
                double min_prev = INF;
                
                // Transitions
                // 1. From same template, previous frame (j-1)
                if (j > 0) {
                    if (current_cost[k][j-1] < min_prev) min_prev = current_cost[k][j-1];
                }
                // 2. From same template, same frame (j) - Self loop (optional, often skipped in simple DTW or penalized)
                // Let's stick to simple Itakura-like: (j-1, t-1), (j, t-1), (j-2, t-1)
                // But here we are iterating t. So we look at current_cost (which is t-1).
                
                if (current_cost[k][j] < min_prev) min_prev = current_cost[k][j]; // Horizontal (j, t-1)
                if (j > 1) {
                     if (current_cost[k][j-2] < min_prev) min_prev = current_cost[k][j-2];
                }

                // 3. Entry from completed word (j=0 only)
                if (j == 0) {
                    // If we are starting the template, we can come from the best score at t-1
                    // But wait, best_score[t-1] represents a completed word at t-1.
                    // So we start matching frame 0 at time t.
                    if (prev_best_score < min_prev) min_prev = prev_best_score;
                }

                next_cost[k][j] = min_prev + dist;
            }

            // Check if this template ended
            double exit_cost = next_cost[k][ref_frames - 1];
            if (exit_cost < best_score[t]) {
                best_score[t] = exit_cost;
                backptr_word[t] = k;
                // We don't strictly know start time here without propagating it, 
                // but for simple reconstruction we can just track the sequence.
                // To track duration, we'd need to propagate 'start_time' through the template.
                // For now, let's just track that a word ended here.
            }
        }

        // Update current_cost for next time step
        for (int k = 0; k < num_templates; k++) {
            for (int j = 0; j < templates[k]->frames; j++) {
                current_cost[k][j] = next_cost[k][j];
            }
        }
        
        // Update prev_best_score for next step (t -> t+1 entry)
        prev_best_score = best_score[t];
    }

    // Backtracking
    // We look for the best ending at the final frame T-1, or nearby?
    // Usually the utterance ends with silence or the last digit.
    // We backtrack from T-1.
    
    // Since we didn't propagate start times, we have to infer the path or use a simpler greedy approach?
    // Actually, without propagating start times (or backpointers within the template), 
    // we can't easily jump back to the start of the word.
    // FIX: We need to propagate the "start time" or "duration" through the template grid.
    
    // Let's re-implement with a simpler structure:
    // Just find the sequence.
    // If we only store "best_score[t]" and "backptr_word[t]", we know WHICH word ended at t.
    // But we don't know WHERE it started, so we don't know where to look for the previous word.
    // We MUST propagate the duration or start time.
}

// Re-implementation with duration propagation
void recognize_corrected(MFCC *test, MFCC **templates, int num_templates) {
    int T = test->frames;
    
    // Global best path to reach time t (completed word)
    double *best_score = (double *)malloc(T * sizeof(double));
    int *word_history = (int *)malloc(T * sizeof(int)); // Which word ended at t
    int *time_history = (int *)malloc(T * sizeof(int)); // Where did the previous word end

    for (int t = 0; t < T; t++) {
        best_score[t] = INF;
        word_history[t] = -1;
        time_history[t] = -1;
    }

    // DP State: [template][frame] -> {cost, start_time}
    // We need two buffers: current (t-1) and next (t)
    typedef struct {
        double cost;
        int start_time;
    } State;

    State **current_state = (State **)malloc(num_templates * sizeof(State *));
    State **next_state = (State **)malloc(num_templates * sizeof(State *));

    for (int k = 0; k < num_templates; k++) {
        int frames = templates[k]->frames;
        current_state[k] = (State *)malloc(frames * sizeof(State));
        next_state[k] = (State *)malloc(frames * sizeof(State));
        for (int j = 0; j < frames; j++) {
            current_state[k][j].cost = INF;
            current_state[k][j].start_time = -1;
        }
    }

    // Initialization
    double prev_global_best = 0.0; // Cost at t=-1
    int prev_global_end = -1;      // End time of previous word

    // Word Insertion Penalty
    double WIP = 0.0; 

    for (int t = 0; t < T; t++) {
        for (int k = 0; k < num_templates; k++) {
            int ref_frames = templates[k]->frames;
            
            for (int j = 0; j < ref_frames; j++) {
                double dist = calculate_distance(test->data[t], templates[k]->data[j], test->dimension);
                
                double min_cost = INF;
                int best_start = -1;

                // 1. From j-1 (Horizontal/Diagonal in grid, but time advances)
                
                // From (t-1, j)
                if (current_state[k][j].cost < min_cost) {
                    min_cost = current_state[k][j].cost;
                    best_start = current_state[k][j].start_time;
                }
                
                // From (t-1, j-1)
                if (j > 0 && current_state[k][j-1].cost < min_cost) {
                    min_cost = current_state[k][j-1].cost;
                    best_start = current_state[k][j-1].start_time;
                }

                // From (t-1, j-2)
                if (j > 1 && current_state[k][j-2].cost < min_cost) {
                    min_cost = current_state[k][j-2].cost;
                    best_start = current_state[k][j-2].start_time;
                }

                // 2. Entry (j=0)
                if (j == 0) {
                    // Can enter from a completed word at t-1
                    if (prev_global_best + WIP < min_cost) {
                        min_cost = prev_global_best + WIP;
                        best_start = t; 
                    }
                }

                next_state[k][j].cost = min_cost + dist;
                next_state[k][j].start_time = best_start;
            }

            // Check exit
            int last = ref_frames - 1;
            if (next_state[k][last].cost < best_score[t]) {
                best_score[t] = next_state[k][last].cost;
                word_history[t] = k;
                time_history[t] = next_state[k][last].start_time - 1;
            }
        }

        // Update buffers
        for (int k = 0; k < num_templates; k++) {
            for (int j = 0; j < templates[k]->frames; j++) {
                current_state[k][j] = next_state[k][j];
            }
        }
        
        prev_global_best = best_score[t];
    }

    // Backtrack
    printf("Recognition Result: ");
    int curr_t = T - 1;

    int result_stack[1000];
    int stack_idx = 0;

    while (curr_t >= 0) {
        int word_idx = word_history[curr_t];
        if (word_idx == -1) break; 
        
        result_stack[stack_idx++] = word_idx;
        curr_t = time_history[curr_t];
    }

    for (int i = stack_idx - 1; i >= 0; i--) {
        printf("%s ", templates[result_stack[i]]->label);
    }
    printf("\n");
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s <test_mfcc>\n", argv[0]);
        return 1;
    }

    // Load templates
    MFCC *templates[MAX_TEMPLATES];
    int num_templates = 0;
    
    char *labels[] = {"1", "2", "3", "4", "5", "6", "7", "8", "9", "o", "z"};
    char filepath[256];

    for (int i = 0; i < 11; i++) {
        sprintf(filepath, "reference_templates_c/%s.mfcc", labels[i]);
        MFCC *t = read_mfcc(filepath, labels[i]);
        if (t) {
            templates[num_templates++] = t;
        } else {
            printf("Warning: Could not load template %s\n", filepath);
        }
    }

    if (num_templates == 0) {
        printf("No templates loaded.\n");
        return 1;
    }

    // Load test file
    MFCC *test = read_mfcc(argv[1], "test");
    if (!test) {
        return 1;
    }

    recognize_corrected(test, templates, num_templates);

    return 0;
}

File: recognizer.py
import numpy as np
import sys
import os

def read_mfcc(filename):
    """Reads MFCC data from a file."""
    try:
        with open(filename, 'r') as f:
            header = f.readline().split()
            if len(header) != 2:
                return None, None
            frames, dimension = int(header[0]), int(header[1])
            data = []
            for _ in range(frames):
                line = f.readline().strip().split()
                data.append([float(x) for x in line])
            data = np.array(data)
            # Cepstral Mean Normalization (CMN)
            mean = np.mean(data, axis=0)
            data = data - mean
            return data, dimension
    except Exception as e:
        print(f"Error reading {filename}: {e}")
        return None, None

def calculate_distance_matrix(test_data, template_data):
    """Calculates Euclidean distance matrix between test and template."""
    # test: (T, D), template: (R, D)
    # dist: (T, R)
    # Using broadcasting: (T, 1, D) - (1, R, D) -> (T, R, D) -> norm -> (T, R)
    diff = test_data[:, np.newaxis, :] - template_data[np.newaxis, :, :]
    dist = np.linalg.norm(diff, axis=2)
    return dist

def level_building(test_data, templates):
    """
    Performs Level Building algorithm.
    test_data: (T, D)
    templates: dict of label -> (R, D)
    """
    T = test_data.shape[0]
    labels = list(templates.keys())
    num_templates = len(labels)
    
    # Parameters
    MAX_LEVELS = 10 # Maximum number of digits expected
    INF = 1e9
    
    # Best score to reach time t at level l
    # best_score[l, t]
    best_score = np.full((MAX_LEVELS + 1, T), INF)
    
    # Backpointers
    # backptr_word[l, t]: index of word ending at t at level l
    backptr_word = np.full((MAX_LEVELS + 1, T), -1, dtype=int)
    # backptr_time[l, t]: start time of word ending at t at level l (actually end of prev word)
    backptr_time = np.full((MAX_LEVELS + 1, T), -1, dtype=int)
    
    # Initialization
    # Level 0: Cost 0 at time -1 (conceptually). 
    # We handle level 1 separately or initialize level 0 boundary.
    # Let's say level 0 ends at time -1 with cost 0.
    
    # To implement this efficiently, we iterate levels.
    # For Level 1: Start at t=0.
    # For Level L: Start at t where Level L-1 ended.
    
    # Pre-compute distance matrices for all templates
    dist_matrices = {}
    for label in labels:
        dist_matrices[label] = calculate_distance_matrix(test_data, templates[label])
        
    # Initialization for Level 1
    # We can start at t=0.
    # We run DTW for each template starting at t=0.
    
    # Actually, standard Level Building computes "best path for template k starting at t-1 and ending at t"
    # But that's expensive (run DTW for every start/end pair).
    # The "One-Pass" equivalent is better:
    # At each time t, we update the state of all templates.
    # But we need to track levels if we want fixed length or just "best sequence".
    # User asked for "continuous detection", usually implies unknown length.
    # So "One-Pass" (Viterbi-like) is appropriate.
    
    # One-Pass State:
    # cost[k, j]: Cost to reach frame j of template k at current time t
    # We also need to track the "history" (best previous word and its end time) for each state?
    # No, usually we just track the "token" that enters the start of the word.
    
    # Let's stick to the One-Pass approach I tried in C, but cleaner.
    
    # Global Best Path to reach time t (end of a word)
    global_best_score = np.full(T, INF)
    global_backptr_word = np.full(T, -1, dtype=int)
    global_backptr_time = np.full(T, -1, dtype=int)
    
    # Active templates state
    # state_cost[k, j]
    state_cost = []
    state_start_time = [] # When did this word start?
    
    for label in labels:
        R = templates[label].shape[0]
        state_cost.append(np.full(R, INF))
        state_start_time.append(np.full(R, -1, dtype=int))
        
    prev_global_best_score = 0.0 # Cost at t=-1
    
    for t in range(T):
        new_state_cost = []
        new_state_start_time = []
        
        # Best score to enter a new word at this time t
        # It comes from the best completed word at t-1
        # Word Insertion Penalty (WIP): Negative value encourages more words
        WIP = -5000.0
        entry_cost = prev_global_best_score + WIP
        entry_start_time = t
        
        current_global_best = INF
        current_best_word = -1
        current_best_time = -1
        
        for k, label in enumerate(labels):
            R = templates[label].shape[0]
            dist = dist_matrices[label][t] # Vector of size R
            
            curr_cost = state_cost[k]
            curr_start = state_start_time[k]
            
            next_cost = np.full(R, INF)
            next_start = np.full(R, -1, dtype=int)
            
            # Vectorized DP update
            # 1. Internal transitions
            # From (j) -> (j) [0]
            # From (j-1) -> (j) [1]
            # From (j-2) -> (j) [2]
            
            # We need to compute min of previous states
            # Construct a padded version for easy shifting
            padded_cost = np.pad(curr_cost, (2, 0), constant_values=INF)
            padded_start = np.pad(curr_start, (2, 0), constant_values=-1)
            
            # Candidates for each j:
            # j (from j in prev): padded_cost[2:]
            # j-1 (from j-1 in prev): padded_cost[1:-1]
            # j-2 (from j-2 in prev): padded_cost[:-2]
            
            c0 = padded_cost[2:]
            c1 = padded_cost[1:-1]
            c2 = padded_cost[:-2]
            
            # Find min cost
            min_prev = np.minimum(c0, np.minimum(c1, c2))
            
            # Determine which one was picked to propagate start_time
            # This is a bit tricky vectorized, let's do argmin
            stacked_costs = np.stack([c0, c1, c2]) # 3 x R
            choices = np.argmin(stacked_costs, axis=0) # R
            
            # Get corresponding start times
            # 0 -> padded_start[2:] (same j)
            # 1 -> padded_start[1:-1] (j-1)
            # 2 -> padded_start[:-2] (j-2)
            
            s0 = padded_start[2:]
            s1 = padded_start[1:-1]
            s2 = padded_start[:-2]
            stacked_starts = np.stack([s0, s1, s2])
            
            # Select best start times
            best_starts = np.choose(choices, stacked_starts)
            
            # 2. Entry transition at j=0
            # If entry_cost < min_prev[0], we take entry
            if entry_cost < min_prev[0]:
                min_prev[0] = entry_cost
                best_starts[0] = entry_start_time
                
            # Update costs
            next_cost = min_prev + dist
            next_start = best_starts
            
            new_state_cost.append(next_cost)
            new_state_start_time.append(next_start)
            
            # Check exit (last frame)
            if next_cost[-1] < current_global_best:
                current_global_best = next_cost[-1]
                current_best_word = k
                current_best_time = next_start[-1] - 1 # End of prev word
                
        state_cost = new_state_cost
        state_start_time = new_state_start_time
        
        global_best_score[t] = current_global_best
        global_backptr_word[t] = current_best_word
        global_backptr_time[t] = current_best_time
        
        prev_global_best_score = current_global_best

    # Backtracking
    result = []
    curr_t = T - 1
    
    # Optional: Find best ending in last few frames
    # best_t = np.argmin(global_best_score[T-5:T]) + (T-5)
    # curr_t = best_t
    
    while curr_t >= 0:
        word_idx = global_backptr_word[curr_t]
        if word_idx == -1:
            break
        result.append(labels[word_idx])
        curr_t = global_backptr_time[curr_t]
        
    return result[::-1]

def main():
    if len(sys.argv) < 2:
        print("Usage: python recognizer.py <test_mfcc>")
        sys.exit(1)
        
    test_file = sys.argv[1]
    
    # Load templates
    templates = {}
    labels = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "o", "z"]
    
    template_dir = "reference_templates_c"
    for l in labels:
        # Try using the _1 variant if available, else fallback
        path = os.path.join(template_dir, f"{l}_1.mfcc")
        if not os.path.exists(path):
             path = os.path.join(template_dir, f"{l}.mfcc")
             
        data, dim = read_mfcc(path)
        if data is not None:
            templates[l] = data
            
    test_data, dim = read_mfcc(test_file)
    if test_data is None:
        print("Error reading test file")
        sys.exit(1)
        
    result = level_building(test_data, templates)
    print("Recognition Result:", " ".join(result))

if __name__ == "__main__":
    main()
